lib/rack/session/redis.rb

```diff
@@ -45,10 +45,45 @@ def find_session(req, sid)
       end
 
       def write_session(req, session_id, new_session, options)
-        with_lock(req, false) do
-          with { |c| c.set session_id, new_session, options }
-          session_id
+        if threadsafe?
+          return transactional_write_session(req, session_id, new_session, options)
         end
+
+        with { |c| c.set session_id, new_session, options }
+        session_id
+      end
+
+      def transactional_write_session(req, session_id, new_session, options)
+        with { |c|
+          # Atomically read the old session and merge it with the new session
+          # Inspired by this: https://stackoverflow.com/a/11311126/7816
+          loop do
+            c.watch session_id
+            old_session = c.get(session_id) || {}
+
+            break if c.multi do |multi|
+              # Using Hash#merge to merge the old and new session is a naïve
+              # strategy which does not always produce the correct result.
+              #
+              # Specifically, if a different request removes a key from the
+              # session in the time between when session was intially read
+              # by the current request and the time when it gets written back,
+              # the deleted keys will reappear.
+              #
+              # Solving this requires determining the intended changes
+              # (insertions/updates/deletions) and selectively applying them,
+              # when writing the session. To accomplish that, we must make a
+              # copy of the initial state of the session at the beginning of
+              # the request, which we can then later compare with the updated
+              # state.
+              #
+              # One might also consider doing a deep merge of the existing and
```

Comment:
is there any reason why you didn't choose this approach?

Reply:
I was worried it would be too large a change. I am happy to give it a shot, though.

Reply:
Do you want me to give it a shot? Does the acceptance of my PR hinge on me doing so?

As I wrote initially, it's not like my changes introduce this problem, they just don't fully solve it. To elaborate a little on why I didn't attempt to fully solve it, I think it's not clear that there is one right way to do it. To users, the session can be thought of as a simply a Hash. To my knowledge, Rack doesn't prescribe any particular structure of this Hash, and so while it might make sense for most users to do a "simple" merge (ie. only consider deletions and insertions at the top level), for others a deep merge might be the best approach. I think the only sensible solution would be to allow users to specify which strategy to use as an option.

Reply:
That would definitely give the most control. I suppose since this doesn't technically depend on `ActiveSupport`, we can't use `#deep_merge` anyway. To answer your question, no, acceptance of your PR does not hinge on this issue. =)